// @oasisprotocol/sapphire-ethers-v6/dist/_cjs/index.cjs
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.wrapEthersProvider = exports.ContractRunnerHasNoProviderError = exports.wrapEthersSigner = exports.SignerHasNoProviderError = exports.NETWORKS = void 0;
const ethers_1 = require("ethers");
const sapphire_paratime_1 = require("@oasisprotocol/sapphire-paratime");
var sapphire_paratime_2 = require("@oasisprotocol/sapphire-paratime");
Object.defineProperty(exports, "NETWORKS", { enumerable: true, get: function () { return sapphire_paratime_2.NETWORKS; } });
function isEthersJsonRpcApiProvider(p) {
    return "send" in p && typeof p.send === "function";
}
function hasRequestFunction(p) {
    return "request" in p && typeof p.request === "function";
}
function getProviderFromContractRunner(r) {
    if (isEthersJsonRpcApiProvider(r)) {
        return r;
    }
    if (r.provider && isEthersJsonRpcApiProvider(r.provider)) {
        return r.provider;
    }
    return;
}
function makeEthereumRequestFnFromProvider(r) {
    if (hasRequestFunction(r)) {
        return r.request;
    }
    const p = getProviderFromContractRunner(r);
    if (!p)
        return;
    return async (request) => {
        const { method, params } = request;
        return p.send(method, params ?? []);
    };
}
const SAPPHIRE_WRAPPED_ETHERS_SIGNER = "#SAPPHIRE_WRAPPED_ETHERS_SIGNER";
function isWrappedSigner(p) {
    return SAPPHIRE_WRAPPED_ETHERS_SIGNER in p;
}
class SignerHasNoProviderError extends Error {
}
exports.SignerHasNoProviderError = SignerHasNoProviderError;
function hookEthersSend(send, options, request) {
    return (async (tx) => {
        if (tx.data) {
            const cipher = await options.fetcher.cipher({ request });
            tx.data = (0, ethers_1.hexlify)(cipher.encryptCall(tx.data));
        }
        return send(tx);
    });
}
function wrapEthersSigner(upstream, options) {
    if (isWrappedSigner(upstream)) {
        return upstream;
    }
    const filled_options = (0, sapphire_paratime_1.fillOptions)(options);
    let signer;
    if (upstream.provider) {
        const provider = wrapEthersProvider(upstream.provider, filled_options);
        try {
            signer = upstream.connect(provider);
        }
        catch (e) {
            if (e.code !== "UNSUPPORTED_OPERATION")
                throw e;
            signer = upstream;
        }
    }
    else {
        signer = upstream;
    }
    const request = makeEthereumRequestFnFromProvider(signer);
    if (!request) {
        throw new SignerHasNoProviderError("Signer must be connected to a provider!");
    }
    return (0, sapphire_paratime_1.makeTaggedProxyObject)(signer, SAPPHIRE_WRAPPED_ETHERS_SIGNER, filled_options, {
        sendTransaction: hookEthersSend(signer.sendTransaction.bind(signer), filled_options, request),
        signTransaction: hookEthersSend(signer.signTransaction.bind(signer), filled_options, request),
        call: hookEthersCall(signer, "call", filled_options, request),
        estimateGas: hookEthersCall(signer, "estimateGas", filled_options, request),
        connect(provider) {
            const wp = signer.connect(provider);
            return wrapEthersSigner(wp, filled_options);
        },
        request,
    });
}
exports.wrapEthersSigner = wrapEthersSigner;
const SAPPHIRE_WRAPPED_ETHERS_PROVIDER = "#SAPPHIRE_WRAPPED_ETHERS_PROVIDER";
function isWrappedProvider(p) {
    return SAPPHIRE_WRAPPED_ETHERS_PROVIDER in p;
}
class ContractRunnerHasNoProviderError extends Error {
}
exports.ContractRunnerHasNoProviderError = ContractRunnerHasNoProviderError;
function wrapEthersProvider(provider, options) {
    if (isWrappedProvider(provider)) {
        return provider;
    }
    const request = makeEthereumRequestFnFromProvider(provider);
    if (!request) {
        throw new Error("Couldn't make request function!");
    }
    const filled_options = (0, sapphire_paratime_1.fillOptions)(options);
    return (0, sapphire_paratime_1.makeTaggedProxyObject)(provider, SAPPHIRE_WRAPPED_ETHERS_PROVIDER, filled_options, {
        call: hookEthersCall(provider, "call", filled_options, request),
        estimateGas: hookEthersCall(provider, "estimateGas", filled_options, request),
        request,
    });
}
exports.wrapEthersProvider = wrapEthersProvider;
function isEthersSigner(upstream) {
    return upstream instanceof ethers_1.AbstractSigner;
}
async function sendUnsignedCall(fn, call, is_already_enveloped, cipher) {
    let call_data = call.data ?? undefined;
    if (!is_already_enveloped) {
        call_data = (0, ethers_1.hexlify)(cipher.encryptCall(call.data));
    }
    return await fn?.({
        ...call,
        data: call_data ? (0, ethers_1.hexlify)(call_data) : "0x",
    });
}
function hookEthersCall(runner, method, options, request) {
    return async (call) => {
        const cipher = await options.fetcher.cipher({ request });
        const is_already_enveloped = (0, sapphire_paratime_1.isCalldataEnveloped)(call.data ?? undefined);
        const f = runner[method];
        if (!f) {
            throw new Error(`${method} not found in runner!`);
        }
        const res = await sendUnsignedCall(f.bind(runner), isEthersSigner(runner) ? await runner.populateCall(call) : call, is_already_enveloped, cipher);
        if (!is_already_enveloped && res && typeof res === "string") {
            return cipher.decryptResult(res);
        }
        return res;
    };
}
